docs/sdg/sdg-flow-yaml.md

```diff
@@ -0,0 +1,164 @@
+# SDG API - Add a file format for defining custom Flows
+
+## Problem Statement
+
+The `instructlab/sdg` library is introducing more extensive data generation pipelines. To enable customization, we should allow users of the library to provide a configuration file which defines a custom pipeline or extends an existing pipeline.
+
+In terms of the API constructs, a Pipeline is created from a sequence of “block configurations” which express how to instantiate and invoke the individual steps (aka blocks) in the pipeline. A Flow construct serves as a template from which a sequence of block configs can be generated.
+
+## Objective
+
+- Library users can specify a custom flow using a well-defined file format.
+- Library users can either use a custom flow standalone, or combine a custom flow with existing flows.
+- The file format and library can evolve substantially without making breaking changes.
+- Incompatible changes can be introduced while retaining support for existing custom flows for a deprecation period.
+
+## Proposal
+
+We will add an API that instantiates a `Pipeline` object from a list of files:
+
+```python
+pipeline = Pipeline.from_flows(ctx, ['mycustomflow.yaml'])
+```
+
+The YAML file format will mirror the API and look like this:
+
+```yaml
```

Comment:
this is essentially reverting some design change/decision we made as per the discussion in https://github.com/instructlab/dev-docs/pull/98: the flow config was originally proposed and implemented as a YAML file with the almost same structure but during the review process it was suggested to be presented as Python codes

i strongly believe this design PR should attribute this properly and probably should be named as something like "reviving YAML config files"

Reply:
I wasn't involved, but from what I can see in the PR, it went like this

The originally proposed design was:

```
with open('run_config.yaml', 'r') as file:
    run_config = yaml.safe_load(file)
cli_sdg = SDG(run_config, client)  # run config has all the variables like num_samples, pipelinesteps etc
generated_samples = cli_sdg.generate()
```

and Russell asked:

> Does this need to be a yaml file?
>
> or is it actually a Python dict passed to the library?
>
> I actually think it would be a nice simplification to not worry about config files at all, and from the library perspective, assume configuration is passed in via data structures.
> 
> How that config is constructed could be a problem of the library consumer. Maybe they hardcode it. maybe they allow a subset to be considered. Some could be driven by CLI args, for example.

So it sounds like Russell anticipated there would be custom pipelines defined by YAML files, but that would be something implemented in the CLI codebase rather than the library.

You can see Russell still saw it that was [when reviewing this PR](https://github.com/instructlab/dev-docs/pull/109#discussion_r1664438995)

> instead of a config file loaded by the sdg library, we could make this config file CLI-specific.

Because he's thinking about where the model serving config will go, and that it probably shouldn't be in a config file format implemented by the SDG library

I responded with:

>In #86, I converted the default flows to YAML ... and it works well I think. And since it works well, it makes sense to me to not have 2 different ways of defining flows, which helps enforce a level of parity between built-in and custom flow capabilities.

and Russell's response was:

> Standardizing on the yaml format everywhere sounds good to me

(i.e. I'm saying "I hacked on this and, in practice, YAML works well for defining pipelines in all cases" and Russell is saying "indeed it does, I see that now")

I'll look for a way to incorporate all of this into the doc :+1:

Reply:
Captured in commit 4425f8d